modules/flowable-ui-modeler/flowable-ui-modeler-logic/src/main/java/org/flowable/app/service/editor/FlowableModelQueryService.java

```diff
@@ -150,121 +158,240 @@ public ResultListDataRepresentation getModelsToIncludeInAppDefinition() {
-                    model.setKey(formkey);
-                    model.setName(formName);
-                    model.setDescription(formDesc);
-                    model.setModelType(AbstractModel.MODEL_TYPE_FORM_RDS);
-
-                    List existingFormModels = modelRepository.findByKeyAndType(formkey, AbstractModel.MODEL_TYPE_FORM_RDS);
-                    if(existingFormModels.size() > 0) {
-                        Model existFormModel = (Model) existingFormModels.get(0);
-                        modelService.saveModel(existFormModel, formDefinition, null,true, "imported", SecurityUtils.getCurrentUserObject());
-                        ModelInfo modelInfo = new ModelInfo(existFormModel.getId(), existFormModel.getName(), formkey);
-                        formKeyMap.put(formkey, modelInfo);
-
-                    } else {
-                        Model formModel = modelService
-                            .createModel(model, formDefinition, SecurityUtils.getCurrentUserObject());
-                        ModelInfo modelInfo = new ModelInfo(formModel.getId(), formModel.getName(), formModel.getKey());
-                        formKeyMap.put(formkey, modelInfo);
-                    }
+    if (process == null)
+    {
+      throw new BpmnFileValidationException(Collections.singletonMap("message", "Main business process is null"));
+    }
 
-                }
+    List existingProcessModels = modelRepository.findByKeyAndType(process.getId(), AbstractModel.MODEL_TYPE_BPMN);
+    if (!overwrite && CollectionUtils.isNotEmpty(existingProcessModels))
+    {
+      validationErrors.put("message", "Process with same key already exists");
+      validationErrors.put("existingProcess", process.getId());
+      throw new BpmnFileValidationException(validationErrors);
+    }
 
-                ObjectNode modelNode = bpmnJsonConverter.convertToJson(bpmnModel, formKeyMap, null);
+    return process;
+  }
+
+  private Map<String, ModelInfo> validateAndSaveForms(Process process, boolean overwrite)
+      throws BpmnFileValidationException
+  {
+    Map<String, ModelInfo> formKeyMap = new HashMap<>();
+    List<String> existingForms = new ArrayList<>();
+    List<ExtensionElement> forms = process.getExtensionElements().get(RdsForms.FORM_ELEMENT);
+
+    if(CollectionUtils.isNotEmpty(forms))
+    {
+      //First find any existing form with same key
+      for (ExtensionElement eeForm : forms)
+      {
+        String formkey = eeForm.getAttributeValue(null, RdsForms.FORM_KEY);
+        List existingFormModels = modelRepository.findByKeyAndType(formkey, AbstractModel.MODEL_TYPE_FORM_RDS);
+        if (CollectionUtils.isNotEmpty(existingFormModels))
+        {
+          existingForms.add(formkey);
+        }
+      }
+
+      // Return validation result if process with same key or any form with same form key exists if overwrite is false
+      if (!overwrite && CollectionUtils.isNotEmpty(existingForms))
```

Comment:
why no longer failing if process key already exists, ie validationResult.size() > 0  ?

Reply:
scrap that ... i see you are failing earlier in that case.   (so much refactoring - got lost ;))